### PR TITLE
Switch to using pytest-variables for user and API credentials. Fixes #166

### LINKED ISSUE #166: Switch to using pytest-variables for credentials
We should move the handling of credentials to pytest-variables. This means changing the example credentials.yml file to variables.json with suitable top level keys and migrating the tests to use the injected variables. We should also add pytest-variables as a dependency.

The variables file would probably look something like this:

```json
{
  "users": {
    "default": {
      "email": "",
      "password": "",
      "name": ""}
  },
  "api": {
    "user": "",
    "key": ""}
  }
}
```

An example of a migrated test would look something like this:

```python
 def test_that_user_can_login_and_logout(self, mozwebqa, variables):
    user = variables['users']['default']
    from pages.login_page import MozTrapLoginPage
    login_pg = MozTrapLoginPage(mozwebqa)
    home_pg = MozTrapHomePage(mozwebqa)
    Assert.false(home_pg.header.is_user_logged_in)
    login_pg.go_to_login_page()
    login_pg.login(user['email'], user['password'])
    Assert.true(home_pg.header.is_user_logged_in)
    Assert.equal(home_pg.header.username_text, user['name'])
    home_pg.header.click_logout()
    Assert.false(home_pg.header.is_user_logged_in)
```

This repository has a `conftest.py` with fixtures. Instead of these using `request.getfuncargvalue('mozwebqa')` they should be able to include the `variables` argument to inject the variables into the method.

You can also see https://github.com/mozilla/sumo-tests/commit/df597f07a1a4b0d72f795bc4dc46a7f387532f29 for an example of migrating to pytest-variables.

### PR DESCRIPTION
**Pull request by davehunt**
There was a bit of a refactor needed here. I separated `mozwebqa_logged_in` into `mozwebqa` and `login`. The latter pulls in the former to do what it needs. This makes the tests a little cleaner, and allows `login` to pull in the credentials from `variables`. I also added an `api` fixture that provides tests and other fixtures with the API client.

The final change I made was to make `element` pull in the `product` fixture and handle the correct order to delete them. This did mean modifying the `_do_delete` to accept a 404 as indicating that the item does not exist and therefore considering the removal a success.

### CODE REVIEW

**Comment by davehunt** [comment]
Pinging @bobsilverberg for review.

File: conftest.py [hunk 1]
```diff
@@ -12,45 +10,40 @@
...
-    mozwebqa = request.getfuncargvalue('mozwebqa')
-    credentials = mozwebqa.credentials['default']
-    request.product = MockProduct()
-    api = MoztrapAPI(credentials['api_user'], credentials['api_key'], mozwebqa.base_url)
-    api.create_product(request.product)
-
-    # This acts like a tearDown, running after each test function
+def product(request, api):
+    """Return a product created via the API."""
+    product = MockProduct()
```
Comment (bobsilverberg) [inline conftest.py]: So, we don't need to store this in the request for it to persist? Where does it get stored? In the fixture itself?

File: pages/base_test.py [hunk 2]
```diff
@@ -88,19 +87,15 @@ def delete_run(self, mozwebqa, run, delete_version=False):
         if delete_version:
             self.delete_version(mozwebqa, version=run['version'])
 
-    def create_suite(self, mozwebqa, product, use_API, status='active', case_list=[], **kwargs):
-        if use_API:
-            credentials = mozwebqa.credentials['default']
+    def create_suite(self, mozwebqa, product, api=None, status='active', case_list=[], **kwargs):
+        if api is not None:
```
Comment (bobsilverberg) [inline pages/base_test.py]: Could this be just `if api`? `api` being an empty string would not be a valid value for `api`.

File: conftest.py [hunk 3]
```diff
@@ -12,45 +10,40 @@
...
-    mozwebqa = request.getfuncargvalue('mozwebqa')
     login_pg = MozTrapLoginPage(mozwebqa)
     login_pg.go_to_login_page()
-    login_pg.login()
+    login_pg.login(user['email'], user['password'])
+
 
-    return mozwebqa
+@pytest.fixture
+def api(request, variables):
```
Comment (bobsilverberg) [inline conftest.py]: +1. This is very nice.

File: tests/test_homepage.py [hunk 4]
```diff
@@ -16,32 +16,27 @@ class TestHomepage(BaseTest):
 
     @pytest.mark.moztrap([3385, 3386])
     @pytest.mark.nondestructive
-    def test_that_user_can_login_and_logout(self, mozwebqa):
+    def test_that_user_can_login_and_logout(self, mozwebqa, variables):
```
Comment (bobsilverberg) [inline tests/test_homepage.py]: Do you think it's worth adding an `existing_user` fixture, so we don't have to interact with `variables` inside the test? Maybe this is the only test that would use it and that's why it's not worthwhile.

File: tests/test_pinning_filters.py [hunk 5]
```diff
@@ -177,13 +177,15 @@ def test_that_pinning_filter_on_product_sets_defaults_in_new_product_version(sel
             u'default product is incorrect')
 
     @pytest.mark.moztrap(5931)
-    def test_that_pinning_filter_persists_for_session(self, mozwebqa_logged_in, product, element):
+    def test_that_pinning_filter_persists_for_session(self, api, mozwebqa, login, product, element, variables):
+        user = variables['users']['default']
```
Comment (bobsilverberg) [inline tests/test_pinning_filters.py]: Here's another place where an `existing_user` fixture could be used.

**Comment by bobsilverberg** [comment]
LGTM @davehunt. Nice work! r+ wc

File: conftest.py (hunk 1 above)
Comment (davehunt) [inline conftest.py]: I'm not entirely sure where/how it's stored, but we don't need to store it ourselves. I made sure to test this too.

File: pages/base_test.py (hunk 2 above)
Comment (davehunt) [inline pages/base_test.py]: `api` should be an instance of `MoztrapAPI`, so I'm not sure I get the 'empty string' part of your comment. We could add type checking, and `if api` would certainly work, but here I'm checking if anything other than the default has been passed, which shows an intention to use the API. We'd fail badly if this was somehow a string/boolean/etc.

File: tests/test_homepage.py (hunk 4 above)
Comment (davehunt) [inline tests/test_homepage.py]: Yes, I think I came up with that for later patches on other projects. I'll bring it here too.

File: tests/test_pinning_filters.py (hunk 5 above)
Comment (davehunt) [inline tests/test_pinning_filters.py]: +1

**Comment by davehunt** [comment]
@bobsilverberg read for another review, adhoc results here: http://webqa-ci-staging1.qa.scl3.mozilla.com:8080/job/moztrap.adhoc/1/

**Comment by bobsilverberg** [comment]
Looks fine now, @davehunt. FTR, if I say `r+ wc` you don't need to ask for another review. It means I am satisfied that if the comments are addressed that everything is fine.

**Comment by davehunt** [comment]
I've updated this so we don't try to delete the product multiple times. If you get a chance @bobsilverberg could you give it another review? Thanks!

**Comment by bobsilverberg** [comment]
LGTM now, @davehunt. I'll let you merge so you can coordinate changing the configs for the Jenkins jobs as well.

**Comment by davehunt** [comment]
Merged in https://github.com/mozilla/moztrap-tests/commit/c5d3eb1f8a7c87586d23b268c56dff0c612adc0e

**Pull request closed without merging**